### ISSUE TITLE
How does Script Lab execute the snippets?

### ISSUE THREAD

**Issue by chengtie** [OfficeDev/script-lab#411]
Hello,

I would like to understand how Script Lab is developed.

If playgrounds such as JsFiddle or plunker execute snippets by `iframe` (which is fairly easy), how does Script Lab run the snippets? how does Script Lab interpret eg, commands with JavaScript API for Office?

Could you explain a little bit or point me to the code that handles this?

Cheers

Tie

**Comment by Zlatkovsky** [comment]
Sounds like a lovely suggestion for an article :-).  Will try to do so in the coming weeks.  Short story is that code is posted from client to server (`src/server/server.ts`), where the TypeScript is compiled to JavaScript, and where a template (`src/server/templates/snippet`) creates a single-page HTML file, while another template (`src/server/templates/runner`) produces an outer wrapper.  The snippet is indeed rendered in an iframe, though there is a bit of a trick to get Office.js to work with it properly.

**Comment by chengtie** [comment]
I see... The part that interests me the most is how the iframe and its wrapper work with `Office.js`, the data from/to Excel, and asynchronous calls...

Looking forward to the article.

Cheers

**Comment by Zlatkovsky** [comment]
In the meantime, I would recommend you look at

- src\server\templates\runner
- src\client\public\runner.ts
- src\client\public\heartbeat.ts

**Comment by chengtie** [comment]
In the meantime, what does the server send back? Does the server send back all the changes (values, format) that should be done to Excel sheets, then it is the client who undertakes the changes to Excel sheets?

**Comment by Zlatkovsky** [comment]
No, I think you're thinking about this a little differently than it actually is.

The server is only there to compile TS to JS (which arguably could be done client-side, but to keep the client lightweight we wanted to do it on the server), and to serve the fully-built HTML page from a different domain (for security's sake).

Actual interaction with Excel is done client-side in the generated snippet.

**Comment by chengtie** [comment]
OK... So the `iframe` is in the client-side, and can handle well the generated snippet containing asynchronous calls and interaction with Excel?

**Comment by Zlatkovsky** [comment]
Correct.  For all intents and purposes, the running snippet is just like a regular HTML+JS+CSS add-in, nothing server-related there.
